examples/src/main/java/com/datastax/oss/driver/examples/mapper/killrvideo/user/CreateUserQueryProvider.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.examples.mapper.killrvideo.user;
+
+import com.datastax.oss.driver.api.core.CqlSession;
+import com.datastax.oss.driver.api.core.cql.BoundStatementBuilder;
+import com.datastax.oss.driver.api.core.cql.PreparedStatement;
+import com.datastax.oss.driver.api.core.cql.ResultSet;
+import com.datastax.oss.driver.api.mapper.MapperContext;
+import com.datastax.oss.driver.api.mapper.entity.EntityHelper;
+import com.datastax.oss.driver.api.mapper.entity.saving.NullSavingStrategy;
+import java.time.Instant;
+import java.util.Objects;
+import java.util.UUID;
+
+/**
+ * Provides the implementation of {@link UserDao#create}.
+ *
+ * <p>Package-private visibility is sufficient, this will be called only from the generated DAO
+ * implementation.
+ */
+class CreateUserQueryProvider {
+
+  private final CqlSession session;
+  private final EntityHelper<User> userHelper;
+  private final EntityHelper<UserCredentials> credentialsHelper;
+  private final PreparedStatement preparedInsertCredentials;
+  private final PreparedStatement preparedInsertUser;
+
+  CreateUserQueryProvider(
+      MapperContext context,
+      EntityHelper<User> userHelper,
+      EntityHelper<UserCredentials> credentialsHelper) {
+
+    this.session = context.getSession();
+
+    this.userHelper = userHelper;
+    this.credentialsHelper = credentialsHelper;
+
+    this.preparedInsertCredentials =
+        session.prepare(credentialsHelper.insert().ifNotExists().asCql());
+    this.preparedInsertUser = session.prepare(userHelper.insert().asCql());
+  }
+
+  boolean create(User user, char[] password) {
+    Objects.requireNonNull(user.getUserid());
+    Objects.requireNonNull(user.getEmail());
+
+    if (!insertCredentialsIfNotExists(user.getEmail(), password, user.getUserid())) {
+      // email already exists
+      return false;
+    }
+
+    if (user.getCreatedDate() == null) {
+      user.setCreatedDate(Instant.now());
+    }
+    insertUser(user);
```

Comment:
I now understand why you didn't use a batch here, but what if the user insertion fails, e.g. because of an `UnavailableException`? It would be wiser to delete the corresponding credentials.

Reply:
I've amended the code to clean up any inserted row if there is an exception.